### ISSUE TITLE
Strategy until_and_while_executing not working properly

### ISSUE THREAD

**Issue by loicboutet** [mhenrixon/sidekiq-unique-jobs#199]
Hi, 

We recently added your gem to our project since we had requirement which are exactly the until_and_while_executing as we understand it. 

However after installing configuring it we were expecting for our job to have a unicity in the queue et while processing. The unicity in the queue is working, and we have only one job with the same argument in the queue. However while processing we have 2 or 3 jobs with the same arguments processing at the same time. 

Any hint to solve this problem? Have you encountered it before? 

Here is our Gemfile.lock regarding sidekiq : 
```

    sidekiq (4.2.2)
      concurrent-ruby (~> 1.0)
      connection_pool (~> 2.2, >= 2.2.0)
      rack-protection (~> 1.5)
      redis (~> 3.2, >= 3.2.1)
    sidekiq-ent (1.3.2)
      algorithms
      sidekiq (~> 4, >= 4.0.2)
      sidekiq-pro (~> 3, >= 3.0.1)
    sidekiq-pro (3.3.3)
      sidekiq (>= 4.0.0)
    sidekiq-unique-jobs (4.0.18)
      sidekiq (>= 2.6)
      thor
```

**Comment by Micka33** [comment]
👍

**Comment by seanhandley** [comment]
👍

**Comment by tandrieu** [comment]
Same issue here

**Comment by fuelen** [comment]
I can confirm this too

**Comment by seanhandley** [comment]
FWIW I worked around this as follows using mutex, in case anyone needs an alternative solution:

```ruby
# app/jobs/my_job.rb
$my_job_mutex = Mutex.new

class MyJob < ApplicationJob
  queue_as :default

  def perform
    return if $my_job_mutex.locked?
    $my_job_mutex.synchronize do
      # Important stuff
    end
  end
end
```

This will mean that any other MyJob that gets enqueued while another is running will simply end without doing anything. See https://ruby-doc.org/core-2.2.0/Mutex.html

**Comment by sbonami** [comment]
Confirmed 👍 . Noticed that retries also cause this issue

**Comment by korstiaan** [comment]
👍

**Comment by mhenrixon** [comment]
This has been somewhat fixed in v5.0.5. It isn't a perfect solution but check the readme for how it currently works: https://github.com/mhenrixon/sidekiq-unique-jobs#while-executing

**Comment by RKushnir** [comment]
@seanhandley I don't think your code works as you expect. Depending on the race condition, the job will either return immediately as you expect, or wait on the `synchronize` until the other worker finishes its job, and then proceed with the job. So it removes the concurrency, but doesn't reliably prevent execution.
You should use `try_lock` instead of `if locked?` & `synchronize`.
